csds-client/client/v3/client.go

```diff
@@ -0,0 +1,388 @@
+			case "trafficdirector.googleapis.com:443":
+				key = "TRAFFICDIRECTOR_GCP_PROJECT_NUMBER"
+			}
+			if projectNum := getValueByKeyFromNodeMatcher(c.nodeMatcher, key); projectNum != "" {
+				c.metadata = metadata.Pairs("x-goog-user-project", projectNum)
+			}
+
+			c.clientConn, err = grpc.Dial(c.opts.Uri, grpc.WithTransportCredentials(creds), grpc.WithPerRPCCredentials(perRPC))
+			if err != nil {
+				return err
+			}
+			return nil
+		default:
+			return errors.New("auto authentication mode for this platform is not supported. Please use jwt_file instead")
+		}
+	default:
+		return errors.New("invalid authn_mode")
+	}
+}
+
+// New creates a new client with v3 api version
+func New(option client.ClientOptions) (*ClientV3, error) {
+	c := &ClientV3{
+		opts: option,
+	}
+	if c.opts.Platform != "gcp" {
+		return nil, fmt.Errorf("%s platform is not supported, list of supported platforms: gcp", c.opts.Platform)
+	}
+
+	if err := c.parseNodeMatcher(); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Run connects the client to the uri and calls doRequest
+func (c *ClientV3) Run() error {
+	if err := c.connWithAuth(); err != nil {
+		return err
+	}
+	defer c.clientConn.Close()
+
+	c.csdsClient = csdspb_v3.NewClientStatusDiscoveryServiceClient(c.clientConn)
+	var ctx context.Context
+	if c.metadata != nil {
+		ctx = metadata.NewOutgoingContext(context.Background(), c.metadata)
+	} else {
+		ctx = context.Background()
+	}
+
+	streamClientStatus, err := c.csdsClient.StreamClientStatus(ctx)
+	if err != nil {
+		return err
+	}
+
+	// run once or run with monitor mode
+	for {
+		if err := c.doRequest(streamClientStatus); err != nil {
+			// timeout error
+			// retry to connect
+			if strings.Contains(err.Error(), "RpcSecurityPolicy") {
+				streamClientStatus, err = c.csdsClient.StreamClientStatus(ctx)
+				if err != nil {
+					return err
+				}
+				continue
+			} else {
+				return err
+			}
+		}
+		if c.opts.MonitorInterval != 0 {
+			time.Sleep(c.opts.MonitorInterval)
+		} else {
+			if err = streamClientStatus.CloseSend(); err != nil {
+				return err
+			}
+			return nil
+		}
+	}
+}
+
+// doRequest sends request and print out the parsed response
+func (c *ClientV3) doRequest(streamClientStatus csdspb_v3.ClientStatusDiscoveryService_StreamClientStatusClient) error {
+
+	req := &csdspb_v3.ClientStatusRequest{NodeMatchers: c.nodeMatcher}
+	if err := streamClientStatus.Send(req); err != nil {
+		return err
+	}
+
+	resp, err := streamClientStatus.Recv()
+	if err != nil && err != io.EOF {
+		return err
+	}
+	// post process response
+	if err := printOutResponse(resp, c.opts); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// parseConfigStatus parses each xds config status to string
+func parseConfigStatus(xdsConfig []*csdspb_v3.PerXdsConfig) []string {
+	var configStatus []string
+	for _, perXdsConfig := range xdsConfig {
+		status := perXdsConfig.GetStatus().String()
+		var xds string
+		if perXdsConfig.GetClusterConfig() != nil {
```

Comment:
missing EDS?

Reply:
EDS can be parsed by this generic function here. For visualization, I added parsing relationship between CDS and EDS.

Reply:
Sorry about that. Just added.